### PR TITLE
Progress in declarative language

### PR DESCRIPTION
**Pull request by ateska**
It breaks unit tests.

### CODE REVIEW

**Comment by ateska** [comment]
image

**Comment by PremyslCerny** [comment]
@ateska Can we keep the list/dict iterations in optimizer and remove the translation to value? At least for now.

**Review by PremyslCerny — APPROVED** [review]
Understandable. :)

**Review by PremyslCerny — APPROVED** [review]

**Comment by PremyslCerny** [comment]
@ateska Mergeable from my side.

File: bspump/declarative/expression/statement/whenexpr.py [hunk 1]
```diff
@@ -24,29 +26,86 @@ class WHEN(Expression):
...
+		self.ItemsNormalized = []
+		self.Attributes = {}
+		self.OutletType = None  # Will be determined in `initialize()`
+		self.Else = VALUE(self.App, value=None)
 
-	def __call__(self, context, event, *args, **kwargs):
-		for branch in self.Items:
-			expr_test = branch.get('test')
-			if expr_test is not None:
+	def set(self, key, value):
```
Comment (ateska) [inline bspump/declarative/expression/statement/whenexpr.py:40]: I don't get this one ... when this is called?
It seems to me to be unnecessary, Expressions are quite immutable objects - they don't change after initialisation.
So why this is needed?

File: bspump/declarative/expression/string/joinexpr.py [hunk 2]
```diff
@@ -1,37 +1,59 @@
...
 		else:
 			self.Miss = arg_miss
 
+	def get_outlet_type(self):
+		return str.__name__
+
+	def consult_inlet_type(self, key, child):
+		return str.__name__
+
+	def set(self, key, value):
```
Comment (ateska) [inline bspump/declarative/expression/string/joinexpr.py:32]: ?

File: bspump/declarative/expression/string/regex.py [hunk 3]
```diff
@@ -84,30 +86,44 @@ def __init__(self, app, *, arg_regex, arg_what, arg_items=None, arg_miss=None, a
 
 		# TODO: Regex flags
 
+	def set(self, key, value):
```
Comment (ateska) [inline bspump/declarative/expression/string/regex.py:89]: ?

File: bspump/declarative/expression/utility/mapexpr.py [hunk 4]
```diff
@@ -36,12 +35,34 @@ def __init__(self, app, *, arg_what, arg_in, arg_else=None):
 			self.Default = arg_else
 
 		self.Map = arg_in
+		self.MapNormalized = dict()
+		self.OutletType = None  # Will be determined in `initialize()`
 
+	def set(self, key, value):
```
Comment (ateska) [inline bspump/declarative/expression/utility/mapexpr.py:41]: ?

File: bspump/declarative/abc.py [hunk 5]
```diff
@@ -52,7 +52,10 @@ def walk(self, parent=None, key=None):
 
 		yield (parent, key, self)
 
-		for key in self.Attributes:
+		# Attributes may change during iteration, which could produce an error
+		current_attributes = self.Attributes.copy()
```
Comment (ateska) [inline bspump/declarative/abc.py:56]: Change of attributes is likely an error. When this happens?

**Review by ateska — COMMENTED** [review]
I have few generic questions here. @PremyslCerny

File: bspump/declarative/expression/statement/whenexpr.py (hunk 1 above)
Comment (PremyslCerny) [reply bspump/declarative/expression/statement/whenexpr.py:40]: @ateska Set is called during optimization process to replace the unoptimzed expression with the optimized one. However, this applies only to expression attributes such as self.Then0, but in self.ItemsOptimized the expression stays the same (unoptimized). Therefore, when self.Then0 is changed, the corresponding item must also be changed in self.ItemsOptimized.

File: bspump/declarative/expression/string/joinexpr.py (hunk 2 above)
Comment (PremyslCerny) [reply bspump/declarative/expression/string/joinexpr.py:32]: @ateska The same as above. Expressions in lists or dicts do not pass optimization.

File: bspump/declarative/expression/string/regex.py (hunk 3 above)
Comment (PremyslCerny) [reply bspump/declarative/expression/string/regex.py:89]: @ateska The same as above. Expressions in lists or dicts do not pass optimization.

File: bspump/declarative/expression/utility/mapexpr.py (hunk 4 above)
Comment (PremyslCerny) [reply bspump/declarative/expression/utility/mapexpr.py:41]: @ateska The same as above. Expressions in lists or dicts do not pass optimization.

File: bspump/declarative/abc.py (hunk 5 above)
Comment (PremyslCerny) [reply bspump/declarative/abc.py:56]: @ateska When in f. e. !WHEN you dynamically add a new attribute such as self.Then0, you change the self.Attributes, which may throw the error.

**Comment by PremyslCerny** [comment]
@ateska Commented. I wonder how !WHEN could have been optimized without the changes I did.

File: bspump/declarative/abc.py (hunk 5 above)
Comment (PremyslCerny) [reply bspump/declarative/abc.py:56]: (walk is also used during initialization as well as optimization)

**Pull request merged**